src/modules/AieraChat/services/ably.ts

```diff
@@ -259,122 +299,171 @@ export const useAbly = (): UseAblyReturn => {
-                                            timestamp: jsonObject.created_at,
-                                            type: ChatMessageType.SOURCES,
-                                        };
-                                        setConfirmation(confirmation); // overwrite
-                                    } else {
-                                        setError('Received source confirmation message without sources');
-                                    }
-                                }
-
-                                // Stop streaming if this is the final partial
-                                if (data.is_final) {
-                                    log('Received final partial:', 'debug');
-                                    log('Stopping partials stream.');
-                                    setIsStreaming(false);
-                                }
-                            } catch (err) {
-                                log(`Error handling message: ${String(err)}`, 'error');
-                                setError(`Error handling message: ${(err as Error).message}`);
-                            }
-                        })
-                        .then(() => {
-                            log(`Subscribed to Ably channel ${channelName}`);
-                            if (!isStreamingRef.current) {
-                                log('Starting to stream partials...');
-                                // Update the streaming status if it's the first partial
-                                setIsStreaming(true);
+            // Check if we're already subscribed to avoid duplicate subscriptions
+            if (globalAblyState.subscribedChannels.has(channelName)) {
+                log(`Channel ${channelName} already subscribed, returning existing channel`);
+                return channel;
+            }
+
+            const messageHandler = (message: Message) => {
+                try {
+                    const data = message.data as AblyEncodedData;
+                    log('Received message from Ably:', 'debug', data);
+
+                    // Decode the base64 string
+                    let decodedData;
+                    try {
+                        decodedData = atob(data.content);
+                    } catch (decodingError) {
+                        log(`Error handling message: ${String(decodingError)}`, 'debug');
+                        return; // ignore message if there's no encoded content
+                    }
+
+                    // Parse the JSON
+                    const jsonObject = JSON.parse(decodedData) as AblyMessageData;
+                    log('Decoded Ably message:', 'log', jsonObject);
+
+                    // Short-circuit if we already processed this message
+                    if (partialKeys.current.includes(jsonObject.created_at)) {
+                        log('Skipping duplicate partial message:', 'log', jsonObject);
+                        return;
```

Comment:
In what kind of scenarios could the same message be processed more than once?

Reply:
I was originally under the impression that the nature of the async lambda conversation chain logic resulted in multiple messages getting published to kafka/ably and parallelization caused duplication. I began investigating this further over these past few days and uncovered a potential issue in the Ably lambda where channels weren't getting closed, so multiple websocket connections were staying open (which I believe caused multiples of the same messages getting published, and therefore consumed on the client). I may end up removing this, but for now it's necessary.